### DISCUSSION TITLE
How do Flat indexes work with frequent insertions? Do I need to rebuild index?  Does milvus keep all data in RAM?

### DISCUSSION THREAD

**Discussion by martinenkoEduard** [milvus-io/milvus#18181] [Q&A and General discussion] [answered]
How do Flat indexes work with frequent insertions? Do I need to rebuild index? 
I am building a face recognition system, and want tot use milvus to find the most similar vectors. 
But the workload will be with constant insertion? 

I have read (but the article was quite old) that milvus uses blocks and uses index only for this particular block (maximum ~500mb if i recall it correctly) .
Do I need to rebuild index if there is almost equal amount of search/insert queries, or index will balance itself automatically (because of chucks?)

Additional question: does milvus keep all the data in RAM? 
Or do I need to specifically tell it to move it to ram?
What is the algorithm for that?

**Comment by martinenkoEduard** [comment]
What does .load() do?
Won't seach work without .load()? 


Can someone recommend a good article or a guide that explains how Milvus works, and why do you need tell mivlust to load?

**Comment by martinenkoEduard** [comment]
Do I need to create index only once?

**Comment by yhmo** [comment, marked as answer]
- Milvus split data into segments, data management is based on segment. By default, each segment size is >= 512MB.
- The load() tell milvus which collection should be loaded into memory(your may have 10 collections, but maybe today you only need search one collection).
- After collection.load() is called, milvus load all segments of the collection into memory of query node. If memory is not enough, an error will return.
- If you continually insert, the insert data is accumulated in a "growing segment", this is an in-memory segment which can be searchable. Once its size exceed 512MB, it will be converted to a normal segment, milvus will automatically load the segment into memory if you have called the load().
- If you have specified an index for a vector field, milvus will assign the index task to index node, index node begin to build index for segment one by one. Once a segment index has been finished, milvus will automatically load the index into memory if you have called load().

So, you only need to call load() once, only call index() once. Milvus will automatically load new segment into memory, automatically build index for new segment, automatically load index into memory to replace vector segment.
